src/vmm/src/devices/virtio/queue.rs

```diff
@@ -242,13 +290,170 @@ impl Queue {
+
+        let size = self.avail_ring_size();
+        let slice = mem
+            .get_slice(self.avail_ring_address, size)
+            .map_err(QueueError::MemoryError)?;
+        slice.bitmap().mark_dirty(0, size);
+        self.avail_ring_ptr = slice.ptr_guard_mut().as_ptr().cast();
+
+        let size = self.used_ring_size();
+        let slice = mem
+            .get_slice(self.used_ring_address, size)
+            .map_err(QueueError::MemoryError)?;
+        slice.bitmap().mark_dirty(0, size);
+        self.used_ring_ptr = slice.ptr_guard_mut().as_ptr().cast();
+
+        Ok(())
+    }
+
+    /// Mark memory used for queue objects as dirty.
+    pub fn mark_memory_dirty<M: GuestMemory>(&self, mem: &M) -> Result<(), QueueError> {
+        let size = self.desc_table_size();
+        let slice = mem
+            .get_slice(self.desc_table_address, size)
+            .map_err(QueueError::MemoryError)?;
+        slice.bitmap().mark_dirty(0, size);
+
+        let size = self.avail_ring_size();
+        let slice = mem
+            .get_slice(self.avail_ring_address, size)
+            .map_err(QueueError::MemoryError)?;
+        slice.bitmap().mark_dirty(0, size);
+
+        let size = self.used_ring_size();
+        let slice = mem
+            .get_slice(self.used_ring_address, size)
+            .map_err(QueueError::MemoryError)?;
+        slice.bitmap().mark_dirty(0, size);
+
+        Ok(())
+    }
+
+    /// Get AvailRing.idx
+    #[inline(always)]
+    pub fn avail_ring_idx_get(&self) -> u16 {
+        // SAFETY: `idx` is 1 u16 away from the start
+        unsafe { self.avail_ring_ptr.add(1).read_volatile() }
+    }
+
+    /// Get element from AvailRing.ring at index
+    /// # Safety
+    /// The `index` parameter should be in 0..queue_size bounds
+    #[inline(always)]
+    pub unsafe fn avail_ring_ring_get(&self, index: usize) -> u16 {
+        // SAFETY: `ring` is 2 u16 away from the start
+        unsafe { self.avail_ring_ptr.add(2).add(index).read_volatile() }
+    }
+
+    /// Get AvailRing.used_event
+    #[inline(always)]
+    pub fn avail_ring_used_event_get(&self) -> u16 {
+        // SAFETY: `used_event` is 2 + self.len u16 away from the start
+        unsafe {
+            self.avail_ring_ptr
+                .add(2_usize.unchecked_add(usize::from(self.size)))
+                .read_volatile()
+        }
+    }
+
+    /// Set UsedRing.idx
+    #[inline(always)]
+    pub fn used_ring_idx_set(&mut self, val: u16) {
+        // SAFETY: `idx` is 1 u16 away from the start
+        unsafe {
+            self.used_ring_ptr
+                .add(std::mem::size_of::<u16>())
+                .cast::<u16>()
+                .write_volatile(val)
+        }
+    }
+
+    /// Get element from UsedRing.ring at index
+    /// # Safety
+    /// The `index` parameter should be in 0..queue_size bounds
+    #[inline(always)]
+    pub unsafe fn used_ring_ring_set(&mut self, index: usize, val: UsedElement) {
```

Comment:
Can you please add kani proofs covering these new functions? E.g. just something simple that `kani::assumes` the safety invariants, to verify that they are indeed sufficient to make the function safe.

Reply:
done in the last commit.